File: agent/lib/cost_events.py
"""Best-effort cost event logging for the agent runtime.

Mirrors lib/observability/cost-events.ts. Never raises — costs are
observability, not correctness; a failed insert can't break a run.
Keep the model price table in sync with the TS file; that file is the
source of truth.
"""

from __future__ import annotations

import logging
from typing import Literal

from db import supabase

log = logging.getLogger(__name__)

Department = Literal[
    "engineering",
    "sales",
    "marketing",
    "design",
    "support",
    "ops_finance",
    "manager",
    "in_process",
]

# Prices last reviewed 2026-05; cross-check provider pricing pages quarterly.
# Source of truth: lib/observability/cost-events.ts. Tuple = (input_per_1m, output_per_1m).
_MODEL_PRICES: dict[str, tuple[float, float]] = {
    "gpt-5":             (1.25, 10.00),
    "gpt-5-mini":        (0.25, 2.00),
    "gpt-5-nano":        (0.05, 0.40),
    "claude-opus-4-7":   (15.00, 75.00),
    "claude-sonnet-4-6": (3.00, 15.00),
    "claude-haiku-4-5":  (0.80, 4.00),
    "gpt-4o-mini":       (0.15, 0.60),
}

_warned_unknown_models: set[str] = set()


def estimate_cost(model: str, input_tokens: int, output_tokens: int) -> float:
    """Estimate USD cost. Unknown model → 0.0 (warn once)."""
    prices = _MODEL_PRICES.get(model)
    if prices is None:
        if model not in _warned_unknown_models:
            _warned_unknown_models.add(model)
            log.warning(
                "cost_events.unknown_model model=%s — add it to _MODEL_PRICES.",
                model,
            )
        return 0.0
    in_per_1m, out_per_1m = prices
    in_cost = (max(0, input_tokens) / 1_000_000) * in_per_1m
    out_cost = (max(0, output_tokens) / 1_000_000) * out_per_1m
    # Round to six decimals — matches the column's numeric(12,6).
    return round(in_cost + out_cost, 6)


async def emit_cost_event(
    *,
    space_id: str,
    department: Department | None,
    model: str,
    input_tokens: int = 0,
    output_tokens: int = 0,
    cost_usd: float | None = None,
    run_id: str | None = None,
    tool_name: str | None = None,
) -> None:
    """Insert one CostEvent row. Best-effort: never raises."""
    try:
        if not space_id or not model:
            log.warning("cost_events.skip reason=missing_space_or_model")
            return

        cost = (
            cost_usd
            if cost_usd is not None
            else estimate_cost(model, input_tokens, output_tokens)
        )

        row: dict[str, object] = {
            "spaceId": space_id,
            "department": department,
            "model": model,
            "inputTokens": int(input_tokens or 0),
            "outputTokens": int(output_tokens or 0),
            "costUsd": float(cost),
        }
        if run_id:
            row["runId"] = run_id
        if tool_name:
            row["toolName"] = tool_name

        db = await supabase()
        await db.table("CostEvent").insert(row).execute()
    except Exception as exc:  # noqa: BLE001
        log.warning("cost_events.insert_failed error=%s", exc)


# Test-only helper for resetting the warned-model cache.
def _reset_warned_models() -> None:
    _warned_unknown_models.clear()

File: lib/observability/cost-events.ts
/**
 * Cost event logging — best-effort writes to the CostEvent table.
 *
 * Callers should never await a failed insert. A cost row failing to land
 * is a logging problem, not a correctness problem; we swallow the error,
 * log a warning, and let the run keep going. Read paths (loadRollup) call
 * the SQL function rollup_cost_by_day and tolerate an empty result.
 *
 * Mirrors agent/lib/cost_events.py. Keep the model price table in sync
 * across the two files — there is no shared source.
 */
import { supabase } from '@/lib/supabase';

export type Department =
  | 'engineering'
  | 'sales'
  | 'marketing'
  | 'design'
  | 'support'
  | 'ops_finance'
  | 'manager'
  | 'in_process';

export interface CostEventInput {
  spaceId: string;
  department?: Department;
  model: string;
  inputTokens?: number;
  outputTokens?: number;
  costUsd?: number;
  runId?: string;
  toolName?: string;
}

export interface RollupRow {
  day: string;            // YYYY-MM-DD
  department: string;
  model: string;
  totalInputTokens: number;
  totalOutputTokens: number;
  totalCostUsd: number;
}

// Prices last reviewed 2026-05; cross-check provider pricing pages quarterly.
// Values are USD per 1M tokens — [inputPer1M, outputPer1M].
export const MODEL_PRICES: Record<string, readonly [number, number]> = {
  'gpt-5':            [1.25, 10.00],
  'gpt-5-mini':       [0.25, 2.00],
  'gpt-5-nano':       [0.05, 0.40],
  'claude-opus-4-7':  [15.00, 75.00],
  'claude-sonnet-4-6':[3.00, 15.00],
  'claude-haiku-4-5': [0.80, 4.00],
  'gpt-4o-mini':      [0.15, 0.60],
};

// Track which unknown models we've already warned about so we don't spam
// the logs. Per-process memory; resets on cold start, which is fine.
const _warnedUnknownModels = new Set<string>();

/**
 * Estimate the USD cost of a call. Returns 0 for unknown models (after
 * logging a one-time warning) so callers can keep going without a price.
 */
export function estimateCost(
  model: string,
  inputTokens: number,
  outputTokens: number,
): number {
  const prices = MODEL_PRICES[model];
  if (!prices) {
    if (!_warnedUnknownModels.has(model)) {
      _warnedUnknownModels.add(model);
      console.warn(
        `[cost-events] Unknown model "${model}" — pricing unavailable. ` +
          `Add it to MODEL_PRICES in lib/observability/cost-events.ts.`,
      );
    }
    return 0;
  }
  const [inPer1M, outPer1M] = prices;
  const inCost = (Math.max(0, inputTokens) / 1_000_000) * inPer1M;
  const outCost = (Math.max(0, outputTokens) / 1_000_000) * outPer1M;
  // Round to six decimal places to match the column's numeric(12,6).
  return Math.round((inCost + outCost) * 1_000_000) / 1_000_000;
}

/**
 * Insert one cost event. Best-effort: never throws.
 *
 * If `costUsd` is omitted, we estimate it from `model` + tokens.
 */
export async function emitCostEvent(input: CostEventInput): Promise<void> {
  try {
    if (!input.spaceId || !input.model) {
      console.warn('[cost-events] Missing spaceId or model; skipping insert.');
      return;
    }

    const inputTokens = input.inputTokens ?? 0;
    const outputTokens = input.outputTokens ?? 0;
    const costUsd =
      input.costUsd ?? estimateCost(input.model, inputTokens, outputTokens);

    const row: Record<string, unknown> = {
      spaceId: input.spaceId,
      department: input.department ?? null,
      model: input.model,
      inputTokens,
      outputTokens,
      costUsd,
    };
    if (input.runId) row.runId = input.runId;
    if (input.toolName) row.toolName = input.toolName;

    const { error } = await supabase.from('CostEvent').insert(row);
    if (error) {
      console.warn(`[cost-events] Insert failed: ${error.message}`);
    }
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    console.warn(`[cost-events] Unexpected error during insert: ${msg}`);
  }
}

/**
 * Load the per-day / per-department / per-model rollup. Returns [] on any
 * error so the dashboard can render the empty state rather than 500.
 */
export async function loadRollup(
  spaceId: string,
  days: number = 30,
): Promise<RollupRow[]> {
  try {
    if (!spaceId) return [];
    const { data, error } = await supabase.rpc('rollup_cost_by_day', {
      p_space_id: spaceId,
      p_days: days,
    });
    if (error) {
      console.warn(`[cost-events] Rollup query failed: ${error.message}`);
      return [];
    }
    const rows = (data ?? []) as Array<{
      day: string;
      department: string | null;
      model: string;
      total_input: number | string;
      total_output: number | string;
      total_cost: number | string;
    }>;
    return rows.map((r) => ({
      day: typeof r.day === 'string' ? r.day.slice(0, 10) : String(r.day).slice(0, 10),
      department: r.department ?? 'in_process',
      model: r.model,
      totalInputTokens: Number(r.total_input ?? 0),
      totalOutputTokens: Number(r.total_output ?? 0),
      totalCostUsd: Number(r.total_cost ?? 0),
    }));
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    console.warn(`[cost-events] Unexpected error during rollup: ${msg}`);
    return [];
  }
}

// Test-only helper to reset the unknown-model warning cache.
export function _resetWarnedModels(): void {
  _warnedUnknownModels.clear();
}
